packages/@lwc/engine-core/src/framework/restrictions.ts

```diff
@@ -268,6 +268,11 @@ function getCustomElementRestrictionsDescriptors(elm: HTMLElement): PropertyDesc
 
 function getComponentRestrictionsDescriptors(): PropertyDescriptorMap {
     assertNotProd(); // this method should never leak to prod
+    if (process.env.NODE_ENV === 'test') {
```

Comment:
I tried this change with the reported test case, the new assertion message is
```
    expect(jest.fn()).not.toBeCalled()

    Expected number of calls: 0
    Received number of calls: 1

    1: {Symbol(@@lockerLiveValue): undefined}

      10 |         document.body.appendChild(element);
      11 |         debugger;
    > 12 |         expect(foo).not.toBeCalled();
         |                         ^
      13 |     });
      14 | });

      at Object.toBeCalled (force-app/main/default/lwc/myComponent/__tests__/index.spec.js:12:25)
```
IMO, the better solution is to provide a serializer for LightningElement component instance via `@lwc/jest-serializer` and have it call the `LightningElement.toString()` method

The background is that jest tries to detect the value type. For DOMElement, it uses the `tagName` property to test if the value is a DOMElement. Hence the error reported by the user. The code is in `pretty-format` package.

Reply:
> However, if we want to make it a bit better, IMO the solution is to provide a serializer for LightningElement component instance via `@lwc/jest-serializer` and have it call the `LightningElement.toString()` method

This makes sense; we can open an issue on `lwc-test` for that.

AIUI, if we did this, then we would not need to add any code to LWC OSS to do special handling for Jest. So yeah, that makes sense to me as an alternative.

Reply:
https://github.com/salesforce/lwc-test/issues/168